components/brave_wallet/browser/simple_hash_client.cc

```diff
@@ -256,6 +263,264 @@ void SimpleHashClient::OnFetchAllNFTsFromSimpleHash(
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    GetNftMetadatasCallback callback) {
+  if (nft_identifiers.size() > kSimpleHashMaxBatchSize) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  GURL url = SimpleHashClient::GetNftsUrl(coin, nft_identifiers);
+  if (!url.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  auto internal_callback = base::BindOnce(
+      &SimpleHashClient::OnGetNftsForMetadatas, weak_ptr_factory_.GetWeakPtr(),
+      coin, std::move(nft_identifiers), std::move(callback));
+
+  api_request_helper_->Request("GET", url, "", "", std::move(internal_callback),
+                               MakeBraveServicesKeyHeader(),
+                               {.auto_retry_on_network_change = true});
+}
+
+void SimpleHashClient::OnGetNftsForMetadatas(
+    mojom::CoinType coin,
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    GetNftMetadatasCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode() ||
+      api_request_result.value_body().is_none()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  // A map of composite token IDs to their metadata
+  std::optional<base::flat_map<std::string, mojom::NftMetadataPtr>> metadatas =
+      ParseMetadatas(api_request_result.TakeBody());
+
+  // For each contract address, create the composite key from the corresponding
+  // chain_id and contract_address at the same index, and look up the metadata
+  // in the map using the case insensitive composite key and add it to the token
+  // urls vector (keeping the original order).
+  std::vector<mojom::NftMetadataPtr> nft_metadatas;
+  for (size_t i = 0; i < nft_identifiers.size(); i++) {
+    std::string composite_key;
+    if (coin == mojom::CoinType::ETH) {
+      composite_key = nft_identifiers[i]->chain_id + "." +
+                      nft_identifiers[i]->contract_address + "." +
+                      nft_identifiers[i]->token_id;
+    } else {
+      composite_key = nft_identifiers[i]->chain_id + "." +
+                      nft_identifiers[i]->contract_address;
+    }
+
+    auto it = metadatas->find(base::ToLowerASCII(composite_key));
+    if (it != metadatas->end()) {
+      nft_metadatas.push_back(std::move(it->second));
+    }
+  }
+
+  std::move(callback).Run(std::move(nft_metadatas));
+}
+
+void SimpleHashClient::GetNfts(
+    mojom::CoinType coin,
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    GetNftsCallback callback) {
+  GURL url = SimpleHashClient::GetNftsUrl(coin, nft_identifiers);
+  if (!url.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  // Erase the first kSimpleHashMaxBatchSize elements from the vectors.
+  // Note: there may be fewer than kSimpleHashMaxBatchSize elements in the
+  // vectors.
+  for (size_t i = 0; i < kSimpleHashMaxBatchSize && !nft_identifiers.empty();
+       i++) {
+    nft_identifiers.erase(nft_identifiers.begin());
+  }
+
```

Comment:
why we need to erase them?

Reply:
Also erasing something from vectors front and moreover doing that in the loop has is very bad performace-wise. Just calc a starting index and in loop to the end move everything to a new vector.